nutanix/resource_nutanix_ndb_database_restore.go

```diff
@@ -0,0 +1,406 @@
+				},
+			},
+			"category": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"parent_database_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"parent_source_database_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"lcm_config":   dataSourceEraLCMConfig(),
+			"time_machine": dataSourceEraTimeMachine(),
+			"dbserver_logical_cluster": {
+				Type:     schema.TypeMap,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"database_nodes":   dataSourceEraDatabaseNodes(),
+			"linked_databases": dataSourceEraLinkedDatabases(),
+		},
+	}
+}
+
+func resourceNutanixNDBDatabaseRestoreCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*Client).Era
+	req := &era.DatabaseRestoreRequest{}
+
+	databaseID := ""
+	if dbID, ok := d.GetOk("database_id"); ok {
+		databaseID = dbID.(string)
+	}
+
+	if snapID, ok := d.GetOk("snapshot_id"); ok {
+		req.SnapshotID = utils.StringPtr(snapID.(string))
+	}
+
+	if latestsnap, ok := d.GetOk("latest_snapshot"); ok {
+		req.LatestSnapshot = utils.StringPtr(latestsnap.(string))
+	}
+
+	if uptime, ok := d.GetOk("user_pitr_timestamp"); ok {
+		req.UserPitrTimestamp = utils.StringPtr(uptime.(string))
+	}
+
+	if timezone, ok := d.GetOk("time_zone"); ok {
+		req.TimeZone = utils.StringPtr(timezone.(string))
+	}
+
+	// getting action arguments
+
+	actargs := []*era.Actionarguments{}
+
+	actargs = append(actargs, &era.Actionarguments{
+		Name:  "sameLocation",
+		Value: "true",
+	})
+
+	req.ActionArguments = actargs
+
+	// call the database restore API
+
+	resp, er := conn.Service.DatabaseRestore(ctx, databaseID, req)
+	if er != nil {
+		return diag.FromErr(er)
+	}
+
+	d.SetId(resp.Entityid)
+
+	// Get Operation ID from response of database restore and poll for the operation to get completed.
+	opID := resp.Operationid
+	if opID == "" {
+		return diag.Errorf("error: operation ID is an empty string")
+	}
+	opReq := era.GetOperationRequest{
+		OperationID: opID,
+	}
+
+	log.Printf("polling for operation with id: %s\n", opID)
+
+	// Poll for operation here - Operation GET Call
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"PENDING"},
+		Target:  []string{"COMPLETED", "FAILED"},
+		Refresh: eraRefresh(ctx, conn, opReq),
+		Timeout: d.Timeout(schema.TimeoutCreate),
+		Delay:   eraDelay,
+	}
+
+	if _, errWaitTask := stateConf.WaitForStateContext(ctx); errWaitTask != nil {
+		return diag.Errorf("error waiting to perform db restore	 (%s) to create: %s", resp.Entityid, errWaitTask)
+	}
+
+	return resourceNutanixNDBDatabaseRestoreRead(ctx, d, meta)
+}
+
+func resourceNutanixNDBDatabaseRestoreRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	c := meta.(*Client).Era
+	if c == nil {
+		return diag.Errorf("era is nil")
+	}
+
+	databaseInstanceID := d.Id()
+
+	resp, err := c.Service.GetDatabaseInstance(ctx, databaseInstanceID)
```

Comment:
Why are we reading database instance here?

Reply:
Does the read tells us that on which snapshot ID or timestamp the database is presently at?

Reply:
@siddharth-kulshrestha maybe the reason is that restore of database is a state change of database itself. @Abhishekism9450 Isnt this method implemented in datasource of database or in databases ? cant we directly use it here.